pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -408,16 +408,16 @@ func (nc *Controller) Run(stopCh <-chan struct{}) {
 					break
 				}
 
-				node := obj.(*v1.Node)
-				hash := hash(node.Name, scheduler.UpdateWorkerSize)
+				nodeName := obj.(string)
+				hash := hash(nodeName, scheduler.UpdateWorkerSize)
 
 				select {
 				case <-stopCh:
-					nc.nodeUpdateQueue.Done(node)
+					nc.nodeUpdateQueue.Done(nodeName)
 					return
-				case nc.nodeUpdateChannels[hash] <- node:
+				case nc.nodeUpdateChannels[hash] <- nodeName:
```

Comment:
What's the reason for making this handling asynchronous? Calling Done prior to actually handling the node in doNoScheduleTaintingPassWorker means the same node can get pushed into the nodeUpdateChannels more than once, right?

Normally, controller just spin up n worker threads to take the next item off a workqueue and process it synchronously. Is there a reason that wouldn't work here, instead of this three level informer->workqueue->channel approach?

Reply:
There maybe a race condition here: 

1. worker get item from queue
1. worker get item from Lister
1. item was updated during `doNoScheduleTaintingPass`

I just go through workqueue just now: we dedup before `Done`; but there maybe a race condition during `doNoScheduleTaintingPass` node was updated after `Lister.Get`, and we need to always check the latest version of Node. Or we dependent on Node's update next time, and call `Done` after `doNoScheduleTaintingPass`

Reply:
> Or we dependent on Node's update next time, and call Done after `doNoScheduleTaintingPass`

We can not dependent on that :( , if node is unreachable, we'll not update status.

Reply:
moving all processing of the node to occur before Done() is called would make this simpler to reason about

Reply:
In this case, I think we still need to use map, e.g. FIFO.

Reply:
After `Lister.Get`, before `doNoScheduleTaintingPass` finished, there maybe node update events. We can not ignore this event, and it's better to only handle the latest event if multiple.

Reply:
if an update is pushed into the store the lister uses, the informer will be notified, and it will enqueue the node name into the workqueue. the workqueue will only dedupe if the node name is already enqueued and waiting, so we will not ignore the event

Reply:
oh, got your point; dirty + processing handled that well; thanks .